crates/transport-http/src/layers/auth.rs

```diff
@@ -113,40 +104,32 @@ where
     }
 
     fn call(&mut self, req: Request<B>) -> Self::Future {
-        let headers = req.headers();
-
-        match self.validate(headers) {
-            true => Box::pin(self.inner.call(req)),
-            false => {
-                // Generate a fresh token from the secret and insert it into the request.
-                match create_token_from_secret(&self.secret) {
-                    Ok(token) => {
-                        let mut req = req.clone();
-
-                        req.headers_mut()
-                            .insert(AUTHORIZATION, HeaderValue::from_str(&token).unwrap());
-                        Box::pin(self.inner.call(req))
-                    }
-                    Err(e) => {
-                        let e = TransportErrorKind::custom(e);
-
-                        Box::pin(async move { Err(e) })
-                    }
+        if self.validate() {
+            Box::pin(self.inner.call(req))
+        } else {
+            // Generate a fresh token from the secret and insert it into the request.
+            match create_token_from_secret(&self.secret) {
+                Ok(token) => {
+                    // Decode the token to get the claims.
+                    let validation = Validation::new(jsonwebtoken::Algorithm::HS256);
+                    let decoding_key = DecodingKey::from_secret(self.secret.as_bytes());
+                    let token_data = decode::<Claims>(&token, &decoding_key, &validation);
+                    self.most_recent_claim = token_data.ok().map(|data| data.claims);
+
+                    let mut req = req;
+                    req.headers_mut().insert(AUTHORIZATION, HeaderValue::from_str(&token).unwrap());
+                    Box::pin(self.inner.call(req))
+                }
+                Err(e) => {
+                    let e = TransportErrorKind::custom(e);
+
+                    Box::pin(async move { Err(e) })
                 }
             }
         }
     }
 }
 
-fn get_bearer_token(headers: &HeaderMap) -> Option<String> {
-    let header = headers.get(AUTHORIZATION)?;
-    let auth: &str = header.to_str().ok()?;
-    let prefix = "Bearer ";
-    let index = auth.find(prefix)?;
-    let token: &str = &auth[index + prefix.len()..];
-    Some(token.into())
-}
-
 fn create_token_from_secret(secret: &JwtSecret) -> Result<String, jsonwebtoken::errors::Error> {
     let token = secret.encode(&Claims {
         iat: (SystemTime::now().duration_since(UNIX_EPOCH).unwrap() + Duration::from_secs(60))
```

Comment:
let's convert this to a function of auth that then also updates the most recent claims